langchain4j-alloydb-pg/src/main/java/dev/langchain4j/store/embedding/alloydb/AlloyDBEmbeddingStore.java

```diff
@@ -37,52 +63,142 @@ public class AlloyDBEmbeddingStore implements EmbeddingStore<TextSegment> {
-     * column to store extra metadata in JSON format.
-     * @param ignoreMetadataColumns (Optional) Column(s) to ignore in
+     * @param ignoreMetadataColumnNames (Optional) Column(s) to ignore in
      * pre-existing tables for a document’s
+     * @param distanceStrategy (Defaults: COSINE_DISTANCE) Distance strategy to
+     * use for vector similarity search
+     * @param k (Defaults: 4) Number of Documents to return from search
+     * @param fetchK (Defaults: 20) Number of Documents to fetch to pass to MMR
+     * algorithm
+     * @param lambdaMult (Defaults: 0.5): Number between 0 and 1 that determines
+     * the degree of diversity among the results with 0 corresponding to maximum
+     * diversity and 1 to minimum diversity
      * @param queryOptions (Optional) QueryOptions class with vector search
      * parameters
      */
-    public AlloyDBEmbeddingStore(AlloyDBEngine engine, String tableName, String schemaName, String contentColumn, String embeddingColumn, String idColumn, List<String> metadataColumns, String metadataJsonColumn, List<String> ignoreMetadataColumns, List<String> queryOptions) {
+    public AlloyDBEmbeddingStore(String tableName, String schemaName, String contentColumn, String embeddingColumn,
+            String idColumn, List<MetadataColumn> metadataColumns, String metadataJsonColumn, List<String> ignoreMetadataColumnNames,
+            DistanceStrategy distanceStrategy, Integer k, Integer fetchK, Double lambdaMult, QueryOptions queryOptions, AlloyDBEngine engine) {
         this.engine = engine;
         this.tableName = tableName;
         this.schemaName = schemaName;
         this.contentColumn = contentColumn;
         this.embeddingColumn = embeddingColumn;
         this.idColumn = idColumn;
-        this.metadataColumns = metadataColumns;
         this.metadataJsonColumn = metadataJsonColumn;
-        this.ignoreMetadataColumns = ignoreMetadataColumns;
+        this.metadataColumns = metadataColumns;
+        this.distanceStrategy = distanceStrategy;
+        this.k = k;
+        this.fetchK = fetchK;
+        this.lambdaMult = lambdaMult;
         this.queryOptions = queryOptions;
+
+        // check columns exist in the table
+        verifyEmbeddingStoreColumns(ignoreMetadataColumnNames);
+    }
+
+    private void verifyEmbeddingStoreColumns(List<String> ignoredColumns) {
+        if (!metadataColumns.isEmpty()) {
+            if (!ignoredColumns.isEmpty()) {
+                throw new IllegalArgumentException("Cannot use both metadataColumns and ignoreMetadataColumns at the same time.");
+            }
+        }
+
+        String query = "SELECT column_name, data_type FROM information_schema.columns WHERE table_name = \"" + tableName + "\" AND table_schema = \"" + schemaName + "\"";
+
+        Map<String, String> allColumns = new HashMap();
+
+        try (ResultSet resultSet = engine.getConnection().createStatement().executeQuery(query)) {
```

Comment:
Can you create a separate resoruce for the connection and statement/query?
```suggestion
        try (Connction conn = engine.getConnection(); ResultSet resultSet = conn.createStatement().executeQuery(query)) {
```